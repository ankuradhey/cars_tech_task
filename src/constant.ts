export const constant = {
    host: "https://auto1-mock-server.herokuapp.com",
    carsUrl: "/api/cars",
    colorsUrl: "/api/colors",
    manufacturersUrl: "/api/manufacturers",
    carDetailsUrl: "/api/cars/{stockNumber}",
};
