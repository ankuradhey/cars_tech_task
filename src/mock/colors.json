{"colors":["red","blue","green","black","yellow","white","silver"]}