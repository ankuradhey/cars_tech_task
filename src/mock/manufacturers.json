{
    "manufacturers": [
        {
            "name": "Audi",
            "models": [
                {
                    "name": "100"
                },
                {
                    "name": "200"
                },
                {
                    "name": "50"
                },
                {
                    "name": "80/90"
                },
                {
                    "name": "A1"
                },
                {
                    "name": "A2"
                },
                {
                    "name": "A3"
                },
                {
                    "name": "A4"
                },
                {
                    "name": "A4 Allroad"
                },
                {
                    "name": "A5"
                },
                {
                    "name": "A6"
                },
                {
                    "name": "A6 Allroad"
                },
                {
                    "name": "A7"
                },
                {
                    "name": "A8"
                },
                {
                    "name": "Allroad"
                },
                {
                    "name": "Cabriolet"
                },
                {
                    "name": "Coupe"
                },
                {
                    "name": "Q2"
                },
                {
                    "name": "Q3"
                },
                {
                    "name": "Q5"
                },
                {
                    "name": "Q7"
                },
                {
                    "name": "Quattro"
                },
                {
                    "name": "R8"
                },
                {
                    "name": "RS 3"
                },
                {
                    "name": "RS 4"
                },
                {
                    "name": "RS 5"
                },
                {
                    "name": "RS 6"
                },
                {
                    "name": "RS 7"
                },
                {
                    "name": "RS Q3"
                },
                {
                    "name": "S1"
                },
                {
                    "name": "S3"
                },
                {
                    "name": "S4"
                },
                {
                    "name": "S5"
                },
                {
                    "name": "S6"
                },
                {
                    "name": "S7"
                },
                {
                    "name": "S8"
                },
                {
                    "name": "SQ5"
                },
                {
                    "name": "SQ7"
                },
                {
                    "name": "TT"
                },
                {
                    "name": "TT Roadster"
                },
                {
                    "name": "V8"
                }
            ]
        },
        {
            "name": "BMW",
            "models": [
                {
                    "name": "1er"
                },
                {
                    "name": "2er"
                },
                {
                    "name": "3er"
                },
                {
                    "name": "4er"
                },
                {
                    "name": "5er"
                },
                {
                    "name": "6er"
                },
                {
                    "name": "7er"
                },
                {
                    "name": "8er"
                },
                {
                    "name": "X1"
                },
                {
                    "name": "X2"
                },
                {
                    "name": "X3"
                },
                {
                    "name": "X4"
                },
                {
                    "name": "X5"
                },
                {
                    "name": "X6"
                },
                {
                    "name": "Z1"
                },
                {
                    "name": "Z3"
                },
                {
                    "name": "Z4"
                },
                {
                    "name": "Z8"
                },
                {
                    "name": "i3"
                },
                {
                    "name": "i8"
                }
            ]
        },
        {
            "name": "Chrysler",
            "models": [
                {
                    "name": "300 C"
                },
                {
                    "name": "300 M"
                },
                {
                    "name": "Crossfire"
                },
                {
                    "name": "Daytona Shelby"
                },
                {
                    "name": "ES"
                },
                {
                    "name": "GS"
                },
                {
                    "name": "GTS"
                },
                {
                    "name": "Grand Voyager"
                },
                {
                    "name": "Le Baron"
                },
                {
                    "name": "Neon"
                },
                {
                    "name": "New Yorker"
                },
                {
                    "name": "PT Cruiser"
                },
                {
                    "name": "Saratoga"
                },
                {
                    "name": "Sebring"
                },
                {
                    "name": "Stratus"
                },
                {
                    "name": "Viper"
                },
                {
                    "name": "Vision"
                },
                {
                    "name": "Voyager"
                }
            ]
        },
        {
            "name": "Dodge",
            "models": [
                {
                    "name": "Avenger"
                },
                {
                    "name": "Caliber"
                },
                {
                    "name": "Journey"
                },
                {
                    "name": "Nitro"
                }
            ]
        },
        {
            "name": "Fiat",
            "models": [
                {
                    "name": "124 Spider"
                },
                {
                    "name": "124 Sport/Spider"
                },
                {
                    "name": "126"
                },
                {
                    "name": "127"
                },
                {
                    "name": "131"
                },
                {
                    "name": "500"
                },
                {
                    "name": "500L"
                },
                {
                    "name": "500X"
                },
                {
                    "name": "850 T/900 T/-E"
                },
                {
                    "name": "Albea"
                },
                {
                    "name": "Argenta"
                },
                {
                    "name": "Barchetta"
                },
                {
                    "name": "Brava"
                },
                {
                    "name": "Bravo"
                },
                {
                    "name": "Cinquecento"
                },
                {
                    "name": "Coupe"
                },
                {
                    "name": "Croma"
                },
                {
                    "name": "Doblo"
                },
                {
                    "name": "Ducato"
                },
                {
                    "name": "Fiorino"
                },
                {
                    "name": "Freemont"
                },
                {
                    "name": "Fullback"
                },
                {
                    "name": "Grande Punto"
                },
                {
                    "name": "Idea"
                },
                {
                    "name": "Linea"
                },
                {
                    "name": "Marea"
                },
                {
                    "name": "Multipla"
                },
                {
                    "name": "Palio"
                },
                {
                    "name": "Panda"
                },
                {
                    "name": "Punto"
                },
                {
                    "name": "Qubo"
                },
                {
                    "name": "Regata"
                },
                {
                    "name": "Ritmo"
                },
                {
                    "name": "Scudo"
                },
                {
                    "name": "Sedici"
                },
                {
                    "name": "Seicento"
                },
                {
                    "name": "Stilo"
                },
                {
                    "name": "Strada"
                },
                {
                    "name": "Talento"
                },
                {
                    "name": "Tempra"
                },
                {
                    "name": "Tipo"
                },
                {
                    "name": "Ulysse"
                },
                {
                    "name": "Uno"
                },
                {
                    "name": "X 1/9"
                }
            ]
        },
        {
            "name": "Mercedes-Benz",
            "models": [
                {
                    "name": "107 Coupe/Roadster"
                },
                {
                    "name": "123"
                },
                {
                    "name": "190/190 E"
                },
                {
                    "name": "207 D - 210 D"
                },
                {
                    "name": "A-Klasse"
                },
                {
                    "name": "B-Klasse"
                },
                {
                    "name": "C-Klasse"
                },
                {
                    "name": "CL-Coupe"
                },
                {
                    "name": "CLA-Klasse"
                },
                {
                    "name": "CLC-Klasse"
                },
                {
                    "name": "CLK-Klasse"
                },
                {
                    "name": "CLS-Klasse"
                },
                {
                    "name": "Citan"
                },
                {
                    "name": "SLC-Klasse"
                },
                {
                    "name": "E-Klasse"
                },
                {
                    "name": "E-Klasse T Modell"
                },
                {
                    "name": "G-Modell"
                },
                {
                    "name": "GL-Klasse"
                },
                {
                    "name": "GLA-Klasse"
                },
                {
                    "name": "GLC-Klasse"
                },
                {
                    "name": "GLE-Klasse"
                },
                {
                    "name": "GLK-Klasse"
                },
                {
                    "name": "GLS-Klasse"
                },
                {
                    "name": "GT AMG"
                },
                {
                    "name": "M-Klasse"
                },
                {
                    "name": "MB 100 D"
                },
                {
                    "name": "R-Klasse"
                },
                {
                    "name": "S-Coupe"
                },
                {
                    "name": "S-Klasse"
                },
                {
                    "name": "SL-Klasse"
                },
                {
                    "name": "SLK-Klasse"
                },
                {
                    "name": "SLR McLaren"
                },
                {
                    "name": "SLS AMG"
                },
                {
                    "name": "Sprinter"
                },
                {
                    "name": "Strich Acht"
                },
                {
                    "name": "Transporter"
                },
                {
                    "name": "V-Klasse"
                },
                {
                    "name": "Vaneo"
                },
                {
                    "name": "Viano"
                },
                {
                    "name": "Vito"
                },
                {
                    "name": "Vito Tourer"
                },
                {
                    "name": "X-Klasse"
                }
            ]
        },
        {
            "name": "Porsche",
            "models": [
                {
                    "name": "718 Boxster"
                },
                {
                    "name": "718 Cayman"
                },
                {
                    "name": "911"
                },
                {
                    "name": "918 Spyder"
                },
                {
                    "name": "924"
                },
                {
                    "name": "928"
                },
                {
                    "name": "944"
                },
                {
                    "name": "968"
                },
                {
                    "name": "Boxster"
                },
                {
                    "name": "Cayenne"
                },
                {
                    "name": "Cayman"
                },
                {
                    "name": "Macan"
                },
                {
                    "name": "Panamera"
                },
                {
                    "name": "Panamera Sport Turismo"
                }
            ]
        },
        {
            "name": "Skoda",
            "models": [
                {
                    "name": "Citigo"
                },
                {
                    "name": "Fabia"
                },
                {
                    "name": "Favorit"
                },
                {
                    "name": "Felicia"
                },
                {
                    "name": "Forman"
                },
                {
                    "name": "Karoq"
                },
                {
                    "name": "Kodiaq"
                },
                {
                    "name": "Octavia"
                },
                {
                    "name": "Pickup"
                },
                {
                    "name": "Rapid"
                },
                {
                    "name": "Roomster"
                },
                {
                    "name": "Skoda 105"
                },
                {
                    "name": "Skoda 120"
                },
                {
                    "name": "Skoda 130"
                },
                {
                    "name": "Superb"
                },
                {
                    "name": "Yeti"
                }
            ]
        },
        {
            "name": "Tesla",
            "models": [
                {
                    "name": "Model S"
                },
                {
                    "name": "Model X"
                },
                {
                    "name": "Roadster"
                }
            ]
        },
        {
            "name": "Volkswagen",
            "models": [
                {
                    "name": "Amarok"
                },
                {
                    "name": "Arteon"
                },
                {
                    "name": "Beetle / New Beetle"
                },
                {
                    "name": "Bora"
                },
                {
                    "name": "CC"
                },
                {
                    "name": "Caddy"
                },
                {
                    "name": "Corrado"
                },
                {
                    "name": "Crafter"
                },
                {
                    "name": "Derby"
                },
                {
                    "name": "Eos"
                },
                {
                    "name": "Fox"
                },
                {
                    "name": "Golf I"
                },
                {
                    "name": "Golf II"
                },
                {
                    "name": "Golf III"
                },
                {
                    "name": "Golf IV"
                },
                {
                    "name": "Golf Plus"
                },
                {
                    "name": "Golf V"
                },
                {
                    "name": "Golf VI"
                },
                {
                    "name": "Golf VII"
                },
                {
                    "name": "Golf VII Sportsvan"
                },
                {
                    "name": "Jetta"
                },
                {
                    "name": "Kaefer 1200"
                },
                {
                    "name": "LT"
                },
                {
                    "name": "Lupo"
                },
                {
                    "name": "Passat"
                },
                {
                    "name": "Phaeton"
                },
                {
                    "name": "Polo"
                },
                {
                    "name": "Santana"
                },
                {
                    "name": "Scirocco"
                },
                {
                    "name": "Scirocco II"
                },
                {
                    "name": "Sharan"
                },
                {
                    "name": "T-Roc"
                },
                {
                    "name": "T3"
                },
                {
                    "name": "T4"
                },
                {
                    "name": "T5"
                },
                {
                    "name": "T6"
                },
                {
                    "name": "Taro"
                },
                {
                    "name": "Tiguan"
                },
                {
                    "name": "Tiguan Allspace"
                },
                {
                    "name": "Touareg"
                },
                {
                    "name": "Touran"
                },
                {
                    "name": "Vento"
                }
            ]
        }
    ]
}