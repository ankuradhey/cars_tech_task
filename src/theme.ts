import { DefaultTheme } from "styled-components";

export const defaultTheme: DefaultTheme = {
    borderRadius: "2px",
    palette: {
        common: {
            black: "#4A4A4A",
            white: "#EDEDED",
        },
        primary: {
            main: "#EA7F28",
            contrastText: "#ffffff",
        },
        secondary: {
            main: "#D37324",
            contrastText: "#ffffff",
        },
        tertiary: {
            main: "#ffffff",
            contrastText: "#EDEDED",
        },
    },
    gutterSize: {
        sm: "8px",
        md: "12px",
        lg: "24px",
    },
};
