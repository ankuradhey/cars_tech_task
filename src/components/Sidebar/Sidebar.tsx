import React, { FC, useState, useEffect } from "react";
import { Card, Form } from "react-bootstrap";
import { constant } from "../../constant";
import { FilterState } from "../../types";
import { Button } from "../Button";
import { Text } from "../Text";

type SidebarProp = {
    setFilter: (props: FilterState) => void;
};

export const Sidebar: FC<SidebarProp> = ({ setFilter }) => {
    const [colors, setColors] = useState([]);
    const [manufacturers, setManufacturers] = useState([]);
    const [formValue, setFormValue] = useState({ color: "", manufacturer: "" });

    const fetchColors = async () => {
        const result = await fetch(constant.colorsUrl);
        const data = await result.json();
        setColors(data.colors);
    };

    const fetchManufacturers = async () => {
        const result = await fetch(constant.manufacturersUrl);
        const data = await result.json();
        setManufacturers(data.manufacturers.map((val) => val.name));
    };

    const setFilterState = (key, val) => {
        setFormValue({ ...formValue, [key]: val });
    };

    const filterClick = () => {
        setFilter(formValue);
    };

    useEffect(() => {
        fetchColors();
        fetchManufacturers();
    }, []);

    return (
        <Card className="margin-bottom-3">
            <Card.Body>
                <Form>
                    <Form.Group>
                        <Form.Label>
                            <Text size="small">Color</Text>
                        </Form.Label>
                        <Form.Control
                            as="select"
                            custom
                            onChange={(e) => setFilterState("color", e.target.value)}
                        >
                            <option value="">All car colors</option>
                            {colors.map((val) => {
                                return <option>{val}</option>;
                            })}
                        </Form.Control>
                    </Form.Group>
                    <Form.Group>
                        <Form.Label>
                            <Text size="small">Manufacturer</Text>
                        </Form.Label>
                        <Form.Control
                            as="select"
                            custom
                            onChange={(e) => setFilterState("manufacturer", e.target.value)}
                        >
                            <option value="">All manufacturers</option>
                            {manufacturers.map((val) => {
                                return <option>{val}</option>;
                            })}
                        </Form.Control>
                    </Form.Group>
                </Form>
                <Button variant="primary" className="float-right" onClick={filterClick}>
                    Filter
                </Button>
            </Card.Body>
        </Card>
    );
};
