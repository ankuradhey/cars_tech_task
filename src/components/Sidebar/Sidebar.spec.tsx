import React from "react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import nock from "nock";
import { render, fireEvent } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { defaultTheme } from "../../theme";
import { Sidebar } from "./Sidebar";
import { constant } from "../../constant";
import * as colorsMock from "../../mock/colors.json";
import * as manufacturersMock from "../../mock/manufacturers.json";

describe("Sidebar Component", () => {
    let history, component, setFilter;
    beforeEach(() => {
        setFilter = jest.fn(() => {});
        history = createMemoryHistory();
        const route = "/";
        history.push(route);

        component = render(
            <ThemeProvider theme={defaultTheme}>
                <Router history={history} initialEntries={["/"]}>
                    <Sidebar setFilter={setFilter} />
                </Router>
            </ThemeProvider>
        );

        nock(`${constant.host}`)
            .defaultReplyHeaders({
                "access-control-allow-origin": "*",
                "access-control-allow-credentials": "true",
            })
            .get(constant.colorsUrl)
            .reply(200, colorsMock);

        nock(`${constant.host}`)
            .defaultReplyHeaders({
                "access-control-allow-origin": "*",
                "access-control-allow-credentials": "true",
            })
            .get(constant.manufacturersUrl)
            .reply(200, manufacturersMock);

        history = createMemoryHistory();
    });

    it("Should render Sidebar component", () => {
        const { container, getByText } = component;
        expect(getByText("Color")).toBeInTheDocument();
        expect(getByText("Manufacturer")).toBeInTheDocument();
        expect(container.firstChild).toMatchSnapshot();
    });

    it("Should have default values in dropdowns", () => {
        const { getByTestId } = component;
        expect(getByTestId("color").value).toBe("");
        expect(getByTestId("manufacturer").value).toBe("");
    });

    it("Should have button filter button that triggers search", () => {
        const { getByTestId } = component;
        expect(getByTestId("filter")).toBeInTheDocument();

        fireEvent.click(
            getByTestId("filter"),
            new MouseEvent("click", {
                bubbles: true,
                cancelable: true,
            })
        );

        expect(setFilter).toHaveBeenCalledWith({ color: "", manufacturer: "" });
    });
});
