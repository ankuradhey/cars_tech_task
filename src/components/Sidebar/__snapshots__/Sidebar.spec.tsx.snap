// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Component Should render Sidebar component 1`] = `
<div
  class="margin-bottom-3 card"
>
  <div
    class="card-body"
  >
    <form
      class=""
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          <div
            class="sc-gsTCUz hhHEdt"
          >
            Color
          </div>
        </label>
        <select
          class="custom-select"
          data-testid="color"
        >
          <option
            value=""
          >
            All car colors
          </option>
        </select>
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
        >
          <div
            class="sc-gsTCUz hhHEdt"
          >
            Manufacturer
          </div>
        </label>
        <select
          class="custom-select"
          data-testid="manufacturer"
        >
          <option
            value=""
          >
            All manufacturers
          </option>
        </select>
      </div>
    </form>
    <button
      class="sc-bdfBwQ hfyqFt float-right"
      data-testid="filter"
    >
      Filter
    </button>
  </div>
</div>
`;
